xayn-ai/src/coi/system.rs

```diff
@@ -203,57 +224,101 @@ impl CoiSystem {
         const DAYS_SCALE: f32 = -0.1;
         let horizon = (self.config.horizon.as_secs_f32() * DAYS_SCALE / SECONDS_PER_DAY).exp();
 
-        cois.iter()
-            .map(|coi| {
-                let CoiStats {
-                    view_count: count,
-                    view_time: time,
-                    last_view: last,
-                    ..
-                } = coi.stats();
-                let count = count as f32 / counts;
-                let time = time.as_secs_f32() / times;
-                let days = (now.duration_since(last).unwrap_or_default().as_secs_f32()
-                    * DAYS_SCALE
-                    / SECONDS_PER_DAY)
-                    .exp();
-                let last = ((horizon - days) / (horizon - 1. - f32::EPSILON)).max(0.);
-                ((count + time) * last).max(0.).min(f32::MAX)
-            })
-            .collect()
+        for coi in cois {
+            let CoiStats {
+                view_count: count,
+                view_time: time,
+                last_view: last,
+                ..
+            } = coi.stats();
+            let count = count as f32 / counts;
+            let time = time.as_secs_f32() / times;
+            let days = (now.duration_since(last).unwrap_or_default().as_secs_f32() * DAYS_SCALE
+                / SECONDS_PER_DAY)
+                .exp();
+            let last = ((horizon - days) / (horizon - 1. - f32::EPSILON)).max(0.);
+            let new_relevance = Relevance(((count + time) * last).max(0.).min(f32::MAX));
+
+            let coi_id = coi.id();
+            let new_relevances = self
+                .config
+                .penalty
+                .iter()
+                .map(|&penalty| {
+                    Relevance(
+                        (f32::from(new_relevance) * penalty)
+                            .max(f32::MIN)
+                            .min(f32::MAX),
+                    )
+                })
+                .collect::<Vec<_>>();
```

Comment:
imo it would look weird to split the body of the loop into another function, because it depends on `counts` & `times` which don't make much sense out of context for a single coi. also testing can be done without too much effort for the whole function. shall i still split it out?